File: Python/DP - Multidimensional/1143. Longest Common Subsequence.py
""" 1143. Longest Common Subsequence
Difficulty: Medium
Submission Link: https://leetcode.com/problems/longest-common-subsequence/submissions/1157635827/ """

class Solution(object):
    def longestCommonSubsequence(self, text1, text2):
        """
        :type text1: str
        :type text2: str
        :rtype: int
        """
        dp = [[0 for _ in range(len(text1)+1)] for _ in range(len(text2) + 1)]
        for i in reversed(range(len(text2))):
            for j in reversed(range(len(text1))):
                if text1[j] == text2[i]:
                    dp[i][j] += dp[i+1][j+1] + 1 
                else:
                    dp[i][j] = max(dp[i+1][j], dp[i][j+1])
        return dp[0][0]

File: C/DP - Multidimensional/1143. Longest Common Subsequence.c
/* 1143. Longest Common Subsequence
Difficulty: Medium
Submission Link: https://leetcode.com/problems/longest-common-subsequence/submissions/1157634218/ */

#include <stdio.h>

int max(int a, int b) {
    return (a >= b) ? a : b;
}

int longestCommonSubsequence(char *text1, char *text2) {
    int m = strlen(text1);
    int n = strlen(text2);

    int dp[m + 1][n + 1];

    for (int i = 0; i <= m; i++) {
        for (int j = 0; j <= n; j++) {
            if (i == 0 || j == 0)
                dp[i][j] = 0;
            else if (text1[i - 1] == text2[j - 1])
                dp[i][j] = dp[i - 1][j - 1] + 1;
            else
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
        }
    }

    return dp[m][n];
}